File: python/data_structures/queue.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.num_elems = 0

    def push(self, val):
        new_item = Node(val)
        if self.num_elems:
            self.last.next = new_item
        else:
            self.first = new_item
        self.last = new_item
        self.num_elems = self.num_elems + 1

    def pop(self):
        if self.num_elems:
            self.first = self.first.next
            self.num_elems = self.num_elems - 1

    def front(self):
        if self.num_elems:
            return self.first.val
        else:
            return None

    def back(self):
        if self.num_elems:
            return self.last.val
        else:
            return None

    def size(self) -> int:
        return self.num_elems

    def empty(self) -> bool:
        if self.num_elems:
            return False
        else:
            return True

if __name__ == "__main__":
    queue = Queue()

    for i in range(0,5):
        queue.push(i)
        assert queue.back() == i
        assert queue.size() == i+1

    i = 0
    while not queue.empty():
        assert queue.front() == i
        i = i+1
        queue.pop()

File: cpp/data_structures/queue.cpp
#include <iostream>

template <typename T>
class Queue {
 private: 
  struct Node {
    Node* next;
    T val;
  };

 public:
  Queue();
  ~Queue();
  bool isEmpty();
  void push(T val);
  void pop();
  int size();
  T front();
  T back();
  
 private:
  Node* first;
  Node* last;
  int num_items;
};

template <typename T>
Queue<T>::Queue() {
  first = nullptr;
  last = nullptr;
  num_items = 0;
}

template <typename T>
Queue<T>::~Queue() {
  while (!isEmpty()) pop();
}

template <typename T>
void Queue<T>::push(T val) {
  Node* new_node = new Node;
  new_node->next = nullptr;
  new_node->val = val;
  if (!isEmpty()) {
    last->next = new_node;
  } else {
    first = new_node;
  }
  last = new_node;
  ++num_items;
}

template <typename T>
void Queue<T>::pop() {
  if (!isEmpty()) {
    Node* target = first;
    first = first->next;
    delete target;
    --num_items;
  }
}

template <typename T>
bool Queue<T>::isEmpty() {
  return (num_items == 0);
}

template <typename T>
T Queue<T>::front() {
  if (!isEmpty()) {
    return first->val;
  } else {
    return T();
  }
}

template <typename T>
T Queue<T>::back() {
  if (!isEmpty()) {
    return last->val;
  } else {
    return T();
  }
}

int main(int argc, char** argv) {
  Queue<int> queue;
  for (int i = 0; i < 5; ++i) {
    queue.push(i);
  }

  while (!queue.isEmpty()) {
    std::cout << "front is: " << queue.front() << std::endl;
    std::cout << "back is: " << queue.back() << std::endl;
    queue.pop();
  }

  std::cout << "front on empty: " << queue.front() << std::endl;
  std::cout << "back on empty: " << queue.back() << std::endl;

  return 0;
}
